File: Java/Merge-Sort.java
public class Main {
    
    public static int[] merge(int[] array, int p, int q, int r) {
        int[] lowHalf = new int[q-p+1];
        int[] highHalf = new int[r-q];

        for (int i = p; i<=q; i++) {
            lowHalf[i-p] = array[i];
        }

        for (int j = q+1; j<=r; j++) {
            highHalf[j-q-1] = array[j];
        }

        int i=0, j=0, k=p;
        while (i<lowHalf.length && j<highHalf.length) {
            if (lowHalf[i] < highHalf[j]) {
                array[k] = lowHalf[i++];
            }
            else {
                array[k] = highHalf[j++];
            }
            k++;
        }

        for ( ; i<lowHalf.length; i++) {
            array[k++] = lowHalf[i];
        }

        for ( ; j<highHalf.length; j++) {
            array[k++] = highHalf[j];
        }
        return array;
    }
    
    public static int[] mergesort(int[] array, int p, int r) {
        if (p < r) {
            int q = (p+r)/2;
            array = mergesort(array, p, q);
            array = mergesort(array, q+1, r);
            array = merge(array, p, q, r);
        }
        return array;
    }

    public static void main(String[] args) {
        int[] array = new int[] {5, 2, 8, 12, 8, 2, 6, 1, -34, 100, -325, 930, 0, 7};
        for (int i = 0; i<array.length; i++) {
            System.out.print(array[i]+" ");
        }
        System.out.println("");
        array = mergesort(array, 0, array.length-1);
        for (int i = 0; i<array.length; i++) {
            System.out.print(array[i]+" "); 
        }
        System.out.println("");
    }
}

File: PHP/Merge-Sort.php
<?php
function merge(&$array, $p, $q, $r) {
    $lowSize = $q-$p+1;
    $highSize = $r-$q;
    $lowHalf = array_fill(0, $lowSize, 0);
    $highHalf = array_fill(0, $highSize, 0);
    $k = $p;
    $i = 0;
    $j = 0;
    for ( ; $k<=$q; $k++) {
        $lowHalf[$k-$p] = $array[$k];
    }

    for ( ; $k<=$r; $k++) {
        $highHalf[$k-$q-1] = $array[$k];
    }
    $k = $p;
    $i = 0;
    $j = 0;
    while ($i < $lowSize && $j < $highSize) {
        if ($lowHalf[$i] < $highHalf[$j]) {
            $array[$k] = $lowHalf[$i++];
        }
        else {
            $array[$k] = $highHalf[$j++];
        }
        $k++;
    }

    for ( ; $i < $lowSize; $i++) {
        $array[$k++] = $lowHalf[$i];
    }

    for ( ; $j < $highSize; $j++) {
        $array[$k++] = $highHalf[$j];
    }
}

function mergesort($array, $p, $r) {
    if ($p<$r) {
        $q = floor(($p+$r)/2);
        $array = mergesort($array, $p, $q);
        $array = mergesort($array, $q+1, $r);
        merge($array, $p, $q, $r);
    }
    return $array;
}

$a = [0, 1, 4, 2, -234, 233, 2];
print_r(mergesort($a, 0, count($a)-1));
?>
